File: hospital/AMarotoChivite-HospitalTest/AMarotoChivite-HospitalTest-Java/src/main/java/HospitalCollections/repositories/PatientListRepository.java
package HospitalCollections.repositories;

import HospitalCollections.base.IHospitalRepository;
import HospitalCollections.models.Patient;
import HospitalCollections.models.PatientComparator;

import java.util.ArrayList;
import java.util.List;

import static HospitalCollections.models.PatientComparator.SortOrder.*;

public class PatientListRepository implements IHospitalRepository {
    private final ArrayList<Patient> dataBase = new ArrayList<>();

    @Override
    public List<Patient> getAll() {
        return dataBase;
    }

    @Override
    public Patient getById(String dniPacient) {

        for (Patient patient : dataBase) {
            if (patient.getDni().equals(dniPacient)) {
                return patient;
            }
        }
        return null;
    }

    @Override
    public Patient save(Patient entity) {
        // Update
        for (int i = 0; i < dataBase.size(); i++) {
            if (dataBase.get(i) != null && dataBase.get(i).getDni().equals(entity.getDni())) {
                dataBase.set(i, entity);
                return entity;
            }
        }
        // Add
        dataBase.add(entity);
        return entity;
    }

    @Override
    public Patient deleteById(String dni) {
        for (int i = 0; i < dataBase.size(); i++) {
            if (dataBase.get(i) != null && dataBase.get(i).getDni().equals(dni)) {
                Patient patientDeleted = dataBase.get(i);
                dataBase.set(i, null);
                return patientDeleted;
            }
        }
        return null;
    }

    @Override
    public boolean isFull() {
        int maxSize = 10;
        return dataBase.size() >= maxSize;
    }

    @Override
    public int getNumPacients() {
        return dataBase.size();
    }

    @Override
    public int[] getNumByType() {
        int countNormal = 0;
        int countUrgent = 0;
        for (Patient patient : dataBase) {
            if (patient.getType().equals(Patient.TypePatient.NORMAL)) {
                countNormal++;
            } else if (patient.getType().equals(Patient.TypePatient.URGENT)) {
                countUrgent++;
            }
        }
        int[] counts = new int[2];
        counts[0] = countNormal;
        counts[1] = countUrgent;
        return counts;
    }

    @Override
    public List<Patient> getAllOrderByEnterDate() {
        dataBase.sort(new PatientComparator(DATE_ENTER));
        return dataBase;
    }

    @Override
    public List<Patient> getAllOrderByName() {
        dataBase.sort(new PatientComparator(NAME));
        return dataBase;
    }

    @Override
    public List<Patient> getAllOrderByType() {
        dataBase.sort(new PatientComparator(TYPE));
        return dataBase;
    }
}

File: hospital/AMarotoChivite-HospitalTest/AMarotoChivite-HospitalTest-Kotlin/src/main/kotlin/HospitalCollections/repositories/PatientListRepository.kt
package HospitalCollections.repositories

import HospitalCollections.base.IHospitalRepository
import HospitalCollections.models.Patient
import HospitalCollections.models.Patient.TypePatient

class PatientListRepository : IHospitalRepository {
    private val dataBase = ArrayList<Patient?>()

    override fun getAll(): List<Patient?> {
        return dataBase
    }

    override fun getById(id: String?): Patient? {
        for (patient in dataBase) {
            if (patient!!.getDni() == id) {
                return patient
            }
        }
        return null
    }

    override fun save(entity: Patient?): Patient? {
        // Update
        for (i in dataBase.indices) {
            if (entity != null) {
                if (dataBase[i] != null && dataBase[i]!!.getDni() == entity.getDni()) {
                    dataBase[i] = entity
                    return entity
                }
            }
        }
        // Add
        dataBase.add(entity)
        return entity
    }

    override fun deleteById(id: String?): Patient? {
        for (i in dataBase.indices) {
            if (dataBase[i] != null && dataBase[i]!!.getDni() == id) {
                val patientDeleted = dataBase[i]
                dataBase[i] = null
                return patientDeleted
            }
        }
        return null
    }

    override fun isFull(): Boolean {
        val maxSize = 10
        return dataBase.size >= maxSize
    }

    override val numPacients: Int
        get() = dataBase.size

    override val numByType: IntArray
        get() {
            var countNormal = 0
            var countUrgent = 0
            for (patient in dataBase) {
                if (patient!!.type == TypePatient.NORMAL) {
                    countNormal++
                } else if (patient.type == TypePatient.URGENT) {
                    countUrgent++
                }
            }
            val counts = IntArray(2)
            counts[0] = countNormal
            counts[1] = countUrgent
            return counts
        }

    override val allOrderByEnterDate: List<Patient?>
        get() {
            // Ordenar sobre la marcha sin comparator, con programación funcional
            return dataBase.sortedWith(compareBy { it!!.getDateEnter() })
        }

    override val allOrderByName: List<Patient?>
        get() {
            // Ordenar sobre la marcha sin comparator, con programación funcional
            return dataBase.sortedWith(compareBy { it!!.getName() })
        }

    override val allOrderByType: List<Patient?>
        get() {
            // Ordenar sobre la marcha sin comparator, con programación funcional
            return dataBase.sortedWith(compareBy { it!!.type })
        }
}